File: app/src/main/java/dev/mainhq/schedules/SearchBus.java
package dev.mainhq.schedules;

import dev.mainhq.schedules.utils.BusListElemsAdapter;
import dev.mainhq.schedules.utils.RecyclerViewItemListener;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.io.Serializable;
import java.lang.String;
import java.util.ArrayList;

//instead of creating a new intent, just redo the list if search done in this activity
public class SearchBus extends AppCompatActivity {
    //processed query
    private Serializable queryResult;
    private Serializable dataType;
    //we tmp removed busInfo hashtable gotten from main
    //bcz doesnt work yet so duplicate code...
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.setContentView(R.layout.search_bus);
        Bundle extras = this.getIntent().getExtras();
        if (extras != null) {
            //this.query = extras.getString("query");
            this.queryResult = extras.getSerializable("STMBusData");
            assert this.queryResult != null;
            if (! (queryResult instanceof ArrayList)) throw new IllegalStateException();
            //Object test = extras.get("STMBusData");
            /*
            String routesFile = "stm_data_info/routes.txt";
            ArrayList<String[]> dataQueried = Parser.setup(this, Parser.DataType.busList, this.query);
            //TODO if null, display "Sorry, no matches made"*/
            if (this.queryResult == null) {
                String empty = "No Matches Found";
                TextView txtview = new TextView(this.getApplicationContext());
            }
            else{
                RecyclerView recyclerView = this.findViewById(R.id.search_recycle_view);
                LinearLayoutManager layoutManager = new LinearLayoutManager(this);
                layoutManager.setOrientation(LinearLayoutManager.VERTICAL);
                recyclerView.setLayoutManager(layoutManager);
                recyclerView.addOnItemTouchListener(new RecyclerViewItemListener(getApplicationContext(), recyclerView, new RecyclerViewItemListener.ClickListener() {
                    @Override
                    public void onClick(View view, int position) {
                        Intent intent = new Intent(getApplicationContext(), ChooseBus.class);
                        ConstraintLayout layout = (ConstraintLayout) view;
                        String busName = ((TextView) layout.getChildAt(0)).getText().toString();
                        String busNum = ((TextView) layout.getChildAt(1)).getText().toString();
                        intent.putExtra("busName", busName);
                        intent.putExtra("busNum", busNum);
                        startActivity(intent);
                    }
                    @Override
                    public void onLongClick(View view, int position) {
                    }
                }));
                //need to improve that code to make it more safe
                recyclerView.setAdapter(new BusListElemsAdapter((ArrayList<String[]>) this.queryResult));
            }
        }
        else Log.d("Query", "None");
        this.onClickBack();
    }

    private void chooseBus(View view){
        TextView bus = (TextView) view.findViewById(R.id.busDir);
        TextView busNum = (TextView) view.findViewById(R.id.busNum);
    }
    private void onClickBack(){
        View view = findViewById(R.id.back_button);
        if (view != null) {
            view.setOnClickListener(v -> {
                finish();
            });
            this.finish();
        }
    }
}

File: app/src/main/java/dev/mainhq/bus2go/SearchBus.kt
package dev.mainhq.bus2go

import android.os.Bundle
import android.view.View
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import dev.mainhq.bus2go.adapters.BusListElemsAdapter
import dev.mainhq.bus2go.utils.FuzzyQuery
import dev.mainhq.bus2go.utils.TransitAgency
import dev.mainhq.bus2go.utils.TransitInfo
import dev.mainhq.bus2go.viewmodels.RoomViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.lang.IllegalStateException

//instead of creating a new intent, just redo the list if search done in this activity
class SearchBus : BaseActivity() {

    private lateinit var roomViewModel: RoomViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.search_bus)
        val extras = intent.extras!!
        val query = extras.getString("query") ?: throw IllegalStateException("There must be a query given to start SearchBus")
        //todo optimisation possible by using the results from mainActivity database query
        roomViewModel = ViewModelProvider(this)[RoomViewModel::class.java]
        lifecycleScope.launch { setup(query) }
        findViewById<View>(R.id.back_button)?.setOnClickListener { finish() }
    }
    
    private suspend fun setup(query : String){
        val jobSTM = lifecycleScope.async {
            val list = roomViewModel.queryStmRoutes(FuzzyQuery(query))
            list.toMutableList().map {
                TransitInfo(it.routeId, it.routeName, null, TransitAgency.STM)
            }
        }
        val jobExo = lifecycleScope.async {
            val list = roomViewModel.queryExoRoutes(FuzzyQuery(query, true))
            list.toMutableList().map {
                TransitInfo(it.routeId.split("-", limit = 2)[1], it.routeName, null, TransitAgency.EXO_OTHER)
            }
        }
        val list = jobSTM.await() + jobExo.await()
        withContext(Dispatchers.Main){
            val recyclerView : RecyclerView = findViewById(R.id.search_recycle_view)
            val layoutManager = LinearLayoutManager(applicationContext)
            recyclerView.adapter = BusListElemsAdapter(list)
            layoutManager.orientation = LinearLayoutManager.VERTICAL
            recyclerView.layoutManager = layoutManager
        }
    }
}

